R/ML.SpeedGLMSGD.R

```diff
@@ -0,0 +1,240 @@
+        
+        
+        #rename colums to standardize depending on the number of columns
+        for (i in (1:(ncol(data)))){
+          if (i == 1){
+            colnames(data)[i] <- "Y"
+          }
+          if (i>1){
+            colnames(data)[i]<-paste("x",i-1, sep = "")
+          }
+        }
+       
+       
+        if(!is.null(m.fit)){
+          result<-private$update_coef(X_mat,Y_vals,coef=m.fit,alpha=private$alpha,loop=private$loop)
+          }
+          else {
+            try(m.fit<-private$read_model(),silent = TRUE)
+            if(!is.null(m.fit)) {
+            ## load model
+            result<-private$update_coef(X_mat,Y_vals,coef=m.fit,alpha=private$alpha,loop=private$loop)
+            }
+           else {
+            ## fit the model
+            model <- private$do.fit(X_mat = data, Y_vals = Y_vals, save_model = FALSE )
+            result <- coef(model)
+            }
+          }
+         
+        if (save_model){
+          private$save_model(model = result)
+           
+        }
+              
+        return(result)
+        
+      },
+
+      do.fit = function (X_mat, Y_vals, save_model = FALSE, coef = NULL) {
+        # If we have not yet fit a model, we are using the first n observations as the training set,
+        # Create dataframe
+         X_mat <- data.frame(X_mat)
+         Y_vals <- data.frame(Y_vals)
+         data <- cbind.data.frame(Y_vals,X_mat)
+        
+         #rename colums to standardize depending on the number of columns
+        for (i in (1:(ncol(data)))){
+          if (i == 1){
+            colnames(data)[i] <- "Y"
+          }
+          if (i>1){
+            colnames(data)[i]<-paste("x",i-1, sep = "")
+          }
+        }
+        
+        
+        x_names=colnames(data[2:ncol(data)])
+        
+        formula <- self$create_formula(x_names)
+      
+        model<-speedglm(formula,data = data, family = binomial(logit))
+      
+        if (save_model){
+        private$save_model(model = coef(model))
+        }
+        
+        return(model)
+      },
+      
+      save_model = function(model) {
+        write.matrix(model, self$get_file_name, sep = "\t")
+      },
+      
+      read_model = function() {
+        as.matrix(read.table(self$get_file_name, as.is = TRUE))
+      },
+      
+      predict_lr = function(X_mat,coef){
+        #prediction of the Y
+        X_mat_mtx=as.matrix(X_mat)
+        intercept_row=cbind(intercept=1,X_mat_mtx)
+        coef_mtx=as.matrix(coef)
+        y_pred=0
+        
+        y_pred <- intercept_row%*% coef_mtx
+        #convert the result into a logit function
+        yhat_return=plogis(y_pred)  
+       
+        return(yhat_return)
+      },
+      
+      gradDescent = function(X_mat, Y_vals, theta, alpha, loop)
+      {
+        m = nrow(X_mat)
+        theta<-as.matrix(theta)
+        X_mat<-as.matrix(X_mat)
+        Y_vals<-as.matrix(Y_vals)
+        
+        while ( loop > 0)
```

Comment:
what would happen if you would set `loop = 1` ? Do we want to bound the values for loop in the constructor?

Reply:
also, you could use 
```R
for(i in 1:loop) {
}
```
Then you won't need the `loop - 1`

Reply:
also, you don't need to pass  it as an argument, you could just use `private$loop` here